feat(appmesh): change VirtualService provider to a union-like class

dfezzie opened the pull request:
Fixes #9490

BREAKING CHANGE: the properties virtualRouter and virtualNode of VirtualServiceProps have been replaced with the union-like class VirtualServiceProvider 
* **appmesh**: the method `addVirtualService` has been removed from `IMesh`

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: bec6a09e153e969f04470b48baa0cf6b40faa559
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ed17d7eb-3d6f-4695-91cd-ee5ea906ddec%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:
```diff
@@ -51,18 +51,11 @@ export interface VirtualServiceBaseProps {
-   * @default - At most one of virtualRouter and virtualNode is allowed.
-   */
-  readonly virtualNode?: IVirtualNode;
+  readonly virtualServiceProvider?: VirtualServiceProvider;
```
2 changes here:

1. Let's move this property to `VirtualServiceProps` below.
2. Let's remove the `mesh` property from `VirtualServiceProps` (no reason to provide it redundantly when a `VirtualRouter` or `VirtualNode` were provided).

dfezzie replied:
1. Do we not need the provider in the base props? How else will we define what type of resource backs the virtual service? 

2. Since the provider is optional, I don't think we can remove the `mesh` property. We still need to know what mesh the VS is created as a part of. We can enforce the mesh properties on the VS and Provider do match.

skinny85 replied:
1. We do not, because the provider will be evident from the resource that the `addService` method will be on. For example, for `VirtualNode`, it will be:

```ts
public addService(props: VirtualServiceBaseProps): IVirtualService: {
  return new VirtualService(this, 'VirtualService', {
    ...props,
    provider: ServiceProvider.virtualNode(this),
  });
}
```

And similarly for `Mesh` and `VirtualRouter`.

2. The `provider` should be required, of course 🙂.

dfezzie replied:
> Putting in "Request changes" to clear this from my ToDo list. @dfezzie make sure to re-request my review when this is ready!

I ended up getting sidetracked and haven't addressed all the feedback quite yet

dfezzie replied:
So I've provided a comment on https://github.com/aws/aws-cdk/issues/9490 that argues we should not have an `addService` method on the `IVirtualNode` and `IVirtualRouter` classes. Will wait on your response before I do anything here.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:
```diff
@@ -135,33 +128,24 @@ export class VirtualService extends cdk.Resource implements IVirtualService {
       virtualServiceName: this.physicalName,
       spec: {
-        provider: this.virtualServiceProvider,
+        provider: provider ? {
+          virtualNode: provider?.virtualNodeProvider,
+          virtualRouter: provider?.virtualRouterProvider,
+        }: undefined,
```
```suggestion
        provider: provider
          ? {
            virtualNode: provider?.virtualNodeProvider,
            virtualRouter: provider?.virtualRouterProvider,
          }
          : undefined,
```

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:
```diff
@@ -211,3 +179,68 @@ export interface VirtualServiceAttributes {
+   *
+   * @default - none
+   */
+  readonly virtualRouterProvider?: CfnVirtualService.VirtualRouterServiceProviderProperty;
```
Let's add a third, required, property here, `mesh: IMesh`.

dfezzie replied:
This is present. May be a stale comment?

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:
```diff
@@ -211,3 +179,68 @@ export interface VirtualServiceAttributes {
+
+  public bind(_construct: cdk.Construct): VirtualServiceProviderConfig {
+    return {
+      virtualNodeProvider: this.virtualNode? {
+        virtualNodeName: this.virtualNode.virtualNodeName,
+      }: undefined,
+      virtualRouterProvider: this.virtualRouter? {
+        virtualRouterName: this.virtualRouter.virtualRouterName,
+      }: undefined,
```
```suggestion
      virtualNodeProvider: this.virtualNode
        ? {
          virtualNodeName: this.virtualNode.virtualNodeName,
        }
        : undefined,
      virtualRouterProvider: this.virtualRouter
        ? {
          virtualRouterName: this.virtualRouter.virtualRouterName,
        }
        : undefined,
```

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:
```diff
@@ -211,3 +179,68 @@ export interface VirtualServiceAttributes {
+  public static virtualRouter(virtualRouter: IVirtualRouter): VirtualServiceProvider {
+    return new VirtualServiceProviderImpl(undefined, virtualRouter);
+  }
+
```
Let's add a third constructor here, `mesh(mesh: IMesh): VirtualServiceProvider`.

dfezzie replied:
What is the implication of this? that there is no provider?

skinny85 replied:
I think so. If you want, you can call the method `none` instead: `none(mes: IMesh): VirtualServiceProvider`.

dfezzie replied:
Thinking on this more, our API spec does not require you to define a provider but the resource would not provide any usability in that state. Providing guardrails here ensures that customers have configured their mesh properly.

skinny85 reviewed (changes requested):
Looks great @dfezzie ! A few comments here, mainly related to this comment that I made: https://github.com/aws/aws-cdk/issues/9490#issuecomment-742749875 .

Please also add `addVirtualService()` methods to both `IVirtualNode` and `IVirtualRouter`.

Also, I'm fairly sure that's the last point that's left of #9490, so I would add `Fixes #9490` to the PR description 🙂.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: b824a4046ecaa8c8a66c91f5be67d98ea1e69d79
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=222b6211-cd48-49dd-939c-39af9091bf6b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: a717a442275abd88569897e6530960d829ebcebc
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8a6fddbf-a925-4b33-be23-f8cc3533e4f3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 reviewed (changes requested):
Putting in "Request changes" to clear this from my ToDo list. @dfezzie make sure to re-request my review when this is ready!

dfezzie commented:
> Putting in "Request changes" to clear this from my ToDo list. @dfezzie make sure to re-request my review when this is ready!

I ended up getting sidetracked and haven't addressed all the feedback quite yet

skinny85 commented:
> > Putting in "Request changes" to clear this from my ToDo list. @dfezzie make sure to re-request my review when this is ready!
> 
> I ended up getting sidetracked and haven't addressed all the feedback quite yet

No worries, it's all good 🙂

dfezzie commented:
Sorry for the long delay. Holidays snuck up on me and then this PR fell off my radar 😅 

I've provided a comment on https://github.com/aws/aws-cdk/issues/9490 about the changes requested. Depending on your response there, we may be able to call this PR good to go.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: c46bdb8aaeafa67665fa73b0ab008b6610ed10a9
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=479b2117-a7bb-4b68-bcb9-c391e86ad3e1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 reviewed (dismissed):
@dfezzie I have no problems with not having `addService()` methods in `IVirtualNode` and `IVirtualRouter`. If that's the case, I think you might want to get rid of the existing `addVirtualService()` method in `IMesh` as well...?

Having said that, I do feel strongly that requiring the user to provide the Mesh twice when creating a Virtual Service with a provider is a mistake (as both a VirtualNode and VirtualRouter already have a Mesh). It's just poor usability.

If you're very opposed to my idea of having a third `none()` VirtualServiceProvider, let's at least make `IMesh` an optional argument when creating a `VirtualService`, and we will not not require it when a `provider` has been given. How does that sound?

dfezzie commented:
I think we should make `provider` a required argument altogether, get rid of the `IMesh` argument and just just get it from the provider. How does that sound?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: db7c5a2efe69e6cb54179c547e030427062f36f4
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c15f8e5a-7124-49a7-9a1b-21fc78f303ea%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: fb2dcda96e592efa2e2571d12d467e30282dd7e0
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d6e76c5a-0ee6-4d90-9826-c00f19212458%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

dfezzie commented:
I've added the `none()` method to the class. Gives the customer flexibility to define a Virtual Service without having to provide a provider out of the gate. 

I'm having issues deploying the integ test to my own account, seeing 

```
1 validation error detected: Value 'local' at 'domainValidationOptions.1.member.validationDomain' failed to satisfy constraint: Member must satisfy regular expression pattern: ^(\*\.)?(((?!-)[A-Za-z0-9-]{0,62}[A-Za-z0-9])\.)+((?!-)[A-Za-z0-9-]{1,62}[A-Za-z0-9])$ (Service: AWSCertificateManager
```

Not sure if there's a bug with the deprecated props in `acm.Certificate` or something weird with my setup. I'll dig into it a bit more tomorrow if the build here doesn't pass

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: cb566fcfed95f6cb6d0560903ab48c59bd4fae2d
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=98f81d92-c0b5-4e93-9e50-4aa52867727e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented:
@dfezzie for deploying the `integ.mesh.ts` integ test, maybe reach out to @alexbrjo ? I think he was the last person to modify them.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 56fb7790c08cb749b1f67a665b3672e1300590ab
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9d18b1f1-6d63-4287-bb30-ebec61b19a10%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

dfezzie commented:
I had to remove ACMPCA and ACM from the integ tests as App Mesh does validate the existence of the resources, and as a result you cannot just import from a fake ARN. I think that should fix the last of the issues with the build.

dfezzie commented on packages/@aws-cdk/aws-appmesh/test/integ.mesh.ts:
```diff
@@ -114,9 +102,6 @@ const node3 = mesh.addVirtualNode('node3', {
       unhealthyThreshold: 2,
     },
   })],
-  backendsDefaultClientPolicy: appmesh.ClientPolicy.fileTrust({
-    certificateChain: 'path-to-certificate',
-  }),
```
Not sure when I removed this. Will add back.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: ae5052e80b0996fd5df22f2fb62a9d0db08c8d6a
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3dc334b0-b93a-4575-8124-5cada3ee6598%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: adcf421e07d345cc7af5e6a28032c7abb2471e83
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=dbdf1a82-ffbf-4127-ad37-19cd4b361fac%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: e43c332d34d1ed82415828aee49a964757559748
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9fc1ad4b-50d9-4164-a9e0-db108dfecbdb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/mesh.ts:
```diff
@@ -89,10 +89,15 @@ abstract class MeshBase extends cdk.Resource implements IMesh {
    * Adds a VirtualService with the given id
    */
-  public addVirtualService(id: string, props: VirtualServiceBaseProps = {}): VirtualService {
+  public addVirtualService(
```
Sorry, I don't understand these changes 😕.

The whole point of having `addVirtualService()` on Mesh is so that you don't have to provide the Mesh again. But with your changes, you can do something like:

```ts
const mesh = new appmesh.Mesh(...);
mesh.addVirtualService({
  virtualServiceProvider.none(mesh),
});
```

which just doesn't make sense to me.

I would revert all of these changes here, and this method should be:

```suggestion
  public addVirtualService(id: string, props: VirtualServiceBaseProps = {}): VirtualService {
    return new VirtualService(this, id, {
      ...props,
      virtualServiceProvider: VirtualServiceProvider.none(this),
    });
  }
```

dfezzie replied:
I guess I was thinking of it more as a utility method, a quick way to add a new Virtual Service to your mesh, similar to how we add other resources like Virtual Nodes. If you cannot specify a provider, then essentially the method is useless as it will never produce a usable Virtual Service. If you created a VS as a placeholder and then went to add a provider you would have to remove it from the `addVirtualService` method into a `new VirtualService(...)` call. How it is here, you could have `addVirtualService('VirtualService')` in one synthesized stack which uses the `none` method as the default. Later, you could come along and add a different provider if you so choose. 

We can remove the method all together if you feel it does not fit well.

skinny85 replied:
> I guess I was thinking of it more as a utility method, a quick way to add a new Virtual Service to your mesh

Right, but if you want to specify _any_ property of the created Service, then it doesn't really save anything, right?

```ts
mesh.addVirtualService('Service', {
  virtualServiceName: 'MyService',
  provider: VirtualServiceProvider.none(mesh),
});
```

Doesn't save anything over

```ts
new VirtualService(this, 'Service', {
  virtualServiceName: 'MyService',
  provider: VirtualServiceProvider.none(mesh),
});
```

I think it doesn't make too much sense to over-emphasize the "add a VirtualService to the Mesh with only default properties" use case - it seems to just lead to an awkward API.

> , similar to how we add other resources like Virtual Nodes.

Notice that all other methods int the AppMesh module like `addVirtualNode()`, `addVirtualGateway()`, etc. do _not_ allow you to specify the main owning resource, like the Mesh, again.

I don't think this inconsistency is worth it. I would either not allow passing `virtualServiceProvider` in `addVirtualService()`, or get rid of this method completely, whichever you prefer (if you go with the later, don't forget to add a BREAKING CHANGE note to the PR description).

dfezzie replied:
For now, let's remove it. You're right that there is no real difference between the two method calls

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:
```diff
@@ -72,16 +63,6 @@ export interface VirtualServiceBaseProps {
-   */
-  readonly mesh: IMesh;
-}
-
```
Like I wrote above in `mesh.ts`, I would revert all of these changes, keep `VirtualServiceBaseProps`, and change `readonly mesh: IMesh` in `VirtualServiceProps` to `readonly virtualServiceProvider: VirtualServiceProvider`.

dfezzie replied:
Done!

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:
```diff
@@ -135,59 +116,39 @@ export class VirtualService extends cdk.Resource implements IVirtualService {
-    }
-    if (props.virtualNode) {
-      this.virtualServiceProvider = this.addVirtualNode(props.virtualNode.virtualNodeName);
+    if (provider.mesh != this.mesh) {
```
This check doesn't make sense to me... the code is

```ts
    this.mesh = provider.mesh;
    if (provider.mesh != this.mesh) {
      // ...
    }
```

So the above `if` will never have its condition be true.

dfezzie replied:
Ah good call. Fixed

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:
```diff
@@ -211,3 +172,100 @@ export interface VirtualServiceAttributes {
+  /**
+   * Mesh the provider belongs to
+   */
+  public abstract readonly mesh: IMesh;
```
Let's not expose this (`VirtualServiceProviderConfig` already contains a `readonly mesh: IMesh` property).

dfezzie replied:
We need to access this in the Virtual Service Constructor class. The `VirtualServiceProps` only has access to the provider class: `readonly virtualServiceProvider: VirtualServiceProvider;` so we need to access the mesh there to properly create the associated cloudformation.

skinny85 replied:
Not sure I understand, sorry.

In the constructor of `VirtualService`, you should call `bind()` on the `virtualServiceProvider` you get, and take the Mesh from there.

There's no point in having 2 separate places where a Mesh is surfaced by.

skinny85 replied:
In fact, this is exactly how the code looks today:

```ts
    const provider = props.virtualServiceProvider?.bind(this);
    this.mesh = provider.mesh;
```

So I'm pretty sure this property is unused.

skinny85 replied:
https://github.com/aws/aws-cdk/pull/11978/files?file-filters%5B%5D=.ts#diff-c7421c5311bdc768c94ddba1234eb002ff0c1d2d4a5e4e5d31724ea762ec7948R130-R131

dfezzie replied:
For some reason I was thinking the bind class was returning an instance of the `VirtualServiceProvider` class 🤦  

Fixed in this revision. Good call

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:
```diff
@@ -211,3 +172,100 @@ export interface VirtualServiceAttributes {
+class VirtualServiceProviderImpl extends VirtualServiceProvider {
+  private readonly virtualNode?: IVirtualNode;
+  private readonly virtualRouter?: IVirtualRouter;
+  public readonly mesh: IMesh;
```
Given the above comment, let's also make this field `private readonly`.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:
```diff
@@ -211,3 +172,100 @@ export interface VirtualServiceAttributes {
+    this.virtualRouter = virtualRouter;
+    const providedMesh = this.virtualNode?.mesh || this.virtualRouter?.mesh || mesh;
+    // Only occurs if VirtualNode or VirtualRouter is not provided during construction
+    if (!providedMesh) {
```
I think this is dead code (this `if` will never have its condition as true).

Let's just do this instead:

```ts
     const providedMesh = this.virtualNode?.mesh ?? this.virtualRouter?.mesh ?? mesh!;
```

dfezzie replied:
Didn't know about the `!`. Neat!

skinny85 commented on packages/@aws-cdk/aws-appmesh/test/test.virtual-service.ts:
```diff
@@ -36,4 +36,23 @@ export = {
     test.equal(virtualService.virtualServiceName, virtualServiceName);
     test.done();
   },
+  'When defining a VirtualServiceProvider': {
```
Empty line before this line.

skinny85 commented on packages/@aws-cdk/aws-appmesh/test/test.virtual-service.ts:
```diff
@@ -36,4 +36,23 @@ export = {
+          virtualServiceName: 'testVS',
+          virtualServiceProvider: appmesh.VirtualServiceProvider.virtualNode(differentMeshVn),
+        }), /Virtual Service provider must belong to mesh/);
+      test.done();
```
Empty line before this line.

skinny85 reviewed (changes requested):
Looks almost perfect @dfezzie! Just a tiny change in `Mesh`, and this will be ready to be shipped.

dfezzie commented:
Addressed a few comments, waiting for a response before making any final updates :+1:

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: e678b9e10d00ca99e705a15c6aef1ce7a9b9957a
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=073e11c6-d095-4739-ad70-088ad752d55c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

dfezzie commented on packages/@aws-cdk/aws-appmesh/test/test.virtual-service.ts:
```diff
@@ -34,6 +36,90 @@ export = {
     test.done();
   },
+
+  'When adding a VirtualService to a mesh': {
```
Moved these methods from the `test.mesh.ts` file. Did not modify them beyond changing `mesh.addVirtualService(...)` to `new appmesh.VirtualService(...)`

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 14aef5e13d2e31e50bd86e42c5926406fe0831ea
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=08926c7f-5f9f-4976-98e3-1e61b9566264%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: a37c6dbcbff658291b6dfd4b645e1107bbd2412c
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a8ee6158-fef6-4cbb-b71b-6342a80311b2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:
```diff
@@ -77,9 +63,9 @@ export interface VirtualServiceBaseProps {
+   * The VirtualNode or VirtualRouter which the VirtualService uses as its provider
    */
-  readonly mesh: IMesh;
+  readonly virtualServiceProvider: VirtualServiceProvider;
```
Now that you've removed `addVirtualService()` from `IMesh`, I'm pretty sure the `VirtualServiceBaseProps` is not used anywhere. If I'm right, get rid of the it in favor of `VirtualServiceProps`.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:
```diff
@@ -135,59 +121,35 @@ export class VirtualService extends cdk.Resource implements IVirtualService {
-    if (props.virtualNode) {
-      this.virtualServiceProvider = this.addVirtualNode(props.virtualNode.virtualNodeName);
-    }
+    const provider = props.virtualServiceProvider?.bind(this);
```
I'm 99% sure `provider` is required here, and so the `?.` can be turned to `.`.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:
```diff
@@ -135,59 +121,35 @@ export class VirtualService extends cdk.Resource implements IVirtualService {
-        provider: this.virtualServiceProvider,
+        provider: provider.virtualNodeProvider || provider.virtualRouterProvider
+          ? {
+            virtualNode: provider?.virtualNodeProvider,
+            virtualRouter: provider?.virtualRouterProvider,
```
Same here - `provider` is never `undefined` here, so no need for the `?.`.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-service.ts:
```diff
@@ -135,59 +121,35 @@ export class VirtualService extends cdk.Resource implements IVirtualService {
       virtualServiceName: this.physicalName,
       spec: {
-        provider: this.virtualServiceProvider,
+        provider: provider.virtualNodeProvider || provider.virtualRouterProvider
```
Gotta say, the name `provider` for this local variable is a little confusing... how about `providerConfig` instead?

skinny85 reviewed (dismissed):
Looks great! Couple of cosmetic changes before we merge this in.

skinny85 reviewed (approved):
Looks great @dfezzie!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: ec39db105d0fc515f5fd0de3a2ec5390086c0a2b
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e6819ce8-8172-49e6-8541-d1290fe2c7e5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: f6afe7bede94b3ca2ed03113f2acf8d1e29bd979
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e2812263-5fd2-4ca5-9358-77fe28d46105%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 575ded52892539191f152911b4d16e4ccb3b9d4d
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=21f5aba2-527d-4935-bdd0-c847e441f6ff%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 93fb6178d2b413fa08d4bb985673acfb6b503ea5
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=36d8e076-b753-4152-be3a-381bcc8da3a6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
